From 0d412c29f5d2991e40637533cbc6094c7e5b2dcc Mon Sep 17 00:00:00 2001
From: Julian Lehrhuber <jl@lerry.de>
Date: Mon, 10 Feb 2020 09:32:58 +0100
Subject: [PATCH 3/3] Auto-accept new devices and share all folders

This helps in a headless scenario, where e.g. a NAS should accept
clients on its own
---
 lib/model/model.go | 32 ++++++++++++++++----------------
 1 file changed, 16 insertions(+), 16 deletions(-)

diff --git a/lib/model/model.go b/lib/model/model.go
index c02ecfd8..81deac82 100644
--- a/lib/model/model.go
+++ b/lib/model/model.go
@@ -2130,23 +2130,23 @@ func (m *model) OnHello(remoteID protocol.DeviceID, addr net.Addr, hello protoco
 
 	cfg, ok := m.cfg.Device(remoteID)
 	if !ok {
-		if err := m.db.AddOrUpdatePendingDevice(remoteID, hello.DeviceName, addr.String()); err != nil {
-			l.Warnf("Failed to persist pending device entry to database: %v", err)
-		}
-		m.evLogger.Log(events.PendingDevicesChanged, map[string][]interface{}{
-			"added": {map[string]string{
-				"deviceID": remoteID.String(),
-				"name":     hello.DeviceName,
-				"address":  addr.String(),
-			}},
-		})
-		// DEPRECATED: Only for backwards compatibility, should be removed.
-		m.evLogger.Log(events.DeviceRejected, map[string]string{
-			"name":    hello.DeviceName,
-			"device":  remoteID.String(),
-			"address": addr.String(),
+		m.cfg.Modify(func(conf *config.Configuration) {
+			// Automatically accept device request
+			conf.SetDevice(config.DeviceConfiguration{
+				DeviceID: remoteID,
+				Name:     hello.DeviceName,
+			})
+			// Share all folders with new device
+			for i := range conf.Folders {
+				conf.Folders[i].Devices = append(conf.Folders[i].Devices, config.FolderDeviceConfiguration{
+					DeviceID: remoteID,
+				})
+			}
 		})
-		return errDeviceUnknown
+		if err := m.cfg.Save(); err != nil {
+			l.Warnln("Saving config:", err)
+		}
+		return nil
 	}
 
 	if cfg.Paused {
-- 
2.32.0

